File: scripts/differential/java/P3T15.java
package app.freerouting.io.specctra;

import app.freerouting.board.BasicBoard;
import app.freerouting.board.FixedState;
import app.freerouting.board.Item;
import app.freerouting.geometry.planar.IntBox;
import app.freerouting.io.BoardReadResult;
import java.io.FileDescriptor;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

/**
 * Plan 3 Task 15 differential driver: reads one Specctra DSN with the real reader and then dumps
 * either the board it built or one of the three writers' output verbatim.
 *
 * <p>Usage: {@code P3T15 <file.dsn> <mode>}, with
 *
 * <ul>
 *   <li>{@code 0} — one line per board item in ascending id order, then the warnings list.
 *   <li>{@code 1} — {@code DsnWriter.write(board, out, stem, false)}, byte for byte.
 *   <li>{@code 2} — {@code SesWriter.write(board, out, stem)}, byte for byte.
 *   <li>{@code 3} — {@code RulesWriter.write(board, out, stem)}, byte for byte.
 *   <li>{@code 4} — the {@code p3t3} token stream, delegated to {@link P3T3} so one driver can
 *       sweep the corpus.
 * </ul>
 *
 * <p>The read is exactly {@code scripts/gen-reference/RefWriter.java}'s:
 * {@code DsnReader.readBoard(in, null, null, designName)} with {@code designName} the input
 * file's name with a trailing {@code .dsn} stripped. That is also what
 * {@code crates/fr-dsn/tests/parity_dsn.rs} does, so a diff here and a parity failure there have
 * the same cause.
 *
 * <p>Unlike {@code p3t3}, this driver goes through the whole parser, so it exercises the
 * hand-rolled {@code nextString}/{@code nextStringList}/{@code nextDouble} bypass path that
 * {@code p3t3}'s pure {@code next_token} loop never reaches (Task 3 review).
 *
 * <p>Compiled against the pinned {@code tools/freerouting-2.3.0.jar} (plan ruling 10), whose
 * package layout predates the clone's renames — hence {@code app.freerouting.board.BasicBoard}
 * rather than {@code board.facade.BasicBoard}, exactly as {@code RefWriter.java} imports it.
 */
public final class P3T15 {

  private P3T15() {}

  public static void main(String[] args) throws Exception {
    if (args.length < 2) {
      System.err.println("usage: P3T15 <file.dsn> <mode 0-4>");
      System.exit(2);
    }
    String path = args[0];
    int mode = Integer.parseInt(args[1]);

    if (mode == 4) {
      P3T3.main(new String[] {path});
      return;
    }

    // `FRLogger` logs to `System.out`, and the reader warns freely; bind this driver's own
    // output to the real stdout and silence `System.out` before FRLogger is loaded, exactly as
    // `P3T3` and the `*Probe.java` golden generators do.
    //
    // `System.err` is deliberately left alone. Java can *throw* out of this driver — reading a
    // board with no `(library …)` scope and then calling `RulesWriter.write` NPEs, see
    // `docs/java-quirks.md` — and the uncaught-exception handler's stack trace on stderr, plus
    // the non-zero exit status, are the only things that tell a caller "Java crashed" apart from
    // "Java disagreed". `sweep-p3t15.sh` redirects stderr to a side file and records both exit
    // codes for exactly that reason. Nothing on stderr can reach the diffed stream.
    PrintStream out =
        new PrintStream(new FileOutputStream(FileDescriptor.out), true, StandardCharsets.UTF_8);
    OutputStream raw = new FileOutputStream(FileDescriptor.out);
    System.setOut(new PrintStream(OutputStream.nullOutputStream()));

    String designName = Path.of(path).getFileName().toString().replaceAll("\\.dsn$", "");

    BoardReadResult result;
    try (InputStream in = new FileInputStream(path)) {
      result = DsnReader.readBoard(in, null, null, designName);
    }

    BasicBoard board;
    List<String> warnings;
    switch (result) {
      case BoardReadResult.Success s -> {
        board = s.board();
        warnings = s.warnings();
      }
      case BoardReadResult.OutlineMissing o -> {
        board = o.board();
        warnings = o.warnings();
      }
      case BoardReadResult.ParseError e -> {
        out.println("RESULT ParseError " + e.location() + " | " + e.detail());
        return;
      }
      default -> {
        out.println("RESULT " + result.getClass().getSimpleName());
        return;
      }
    }
    if (board == null) {
      out.println("RESULT NoBoard");
      return;
    }

    switch (mode) {
      case 0 -> dumpItems(out, board, warnings);
      case 1 -> {
        DsnWriter.write(board, raw, designName, false);
        raw.flush();
      }
      case 2 -> {
        SesWriter.write(board, raw, designName);
        raw.flush();
      }
      case 3 -> {
        RulesWriter.write(board, raw, designName);
        raw.flush();
      }
      default -> {
        System.err.println("unknown mode: " + mode);
        System.exit(2);
      }
    }
  }

  /**
   * One line per item: {@code item <id> <kind> layers=<first>..<last> nets=[..] cl=<class>
   * fixed=<state> cmp=<component> bbox=(llx,lly,urx,ury) tiles=<count>}, in ascending id order,
   * then {@code itemcount} and the warnings.
   */
  private static void dumpItems(PrintStream out, BasicBoard board, List<String> warnings) {
    out.println("layers " + board.get_layer_count());
    List<Item> items = new ArrayList<>(board.get_items());
    items.sort(Comparator.comparingInt(Item::get_id_no));
    for (Item it : items) {
      out.println("item " + it.get_id_no() + " " + describe(it));
    }
    out.println("itemcount " + items.size());
    for (String w : warnings) {
      out.println("warning " + w);
    }
  }

  private static String describe(Item it) {
    StringBuilder sb = new StringBuilder(it.getClass().getSimpleName());
    sb.append(" layers=").append(it.first_layer()).append("..").append(it.last_layer());
    sb.append(" nets=[");
    for (int i = 0; i < it.net_count(); i++) {
      sb.append(it.get_net_no(i)).append(',');
    }
    sb.append(']');
    sb.append(" cl=").append(it.clearance_class_no());
    FixedState fixed = it.get_fixed_state();
    sb.append(" fixed=").append(fixed == null ? "null" : fixed.name());
    sb.append(" cmp=").append(it.get_component_no());
    IntBox box = it.bounding_box();
    sb.append(" bbox=(")
        .append(box.ll.x)
        .append(',')
        .append(box.ll.y)
        .append(',')
        .append(box.ur.x)
        .append(',')
        .append(box.ur.y)
        .append(')');
    sb.append(" tiles=").append(it.tile_shape_count());
    return sb.toString();
  }
}

File: scripts/differential/rust/src/bin/p3t15.rs
//! Rust twin of `scripts/differential/java/P3T15.java` (Plan 3 Task 15).
//!
//! Reads one Specctra DSN through `copper-dsn` and dumps either the board it built (mode 0) or one
//! of the three writers' output verbatim (modes 1-3); mode 4 is the `p3t3` token stream, so a
//! single driver can sweep the corpus. See `scripts/differential/README.md`.
//!
//! Because modes 0-3 go through the whole parser rather than the bare scanner, this driver
//! exercises the hand-rolled `next_string`/`next_string_list`/`next_double` bypass path that
//! `p3t3`'s pure `next_token` loop never reaches.

use std::io::{BufWriter, Write};

use copper_board::board::Board;
use copper_board::items::Item;
use copper_dsn::parser::scope_parameter::DsnReadOptions;
use copper_dsn::BoardReadResult;

#[path = "../token_dump.rs"]
mod token_dump;

fn main() {
    let mut args = std::env::args().skip(1);
    let (path, mode) = match (args.next(), args.next()) {
        (Some(path), Some(mode)) => (path, mode.parse::<u8>().expect("mode must be 0-4")),
        _ => {
            eprintln!("usage: p3t15 <file.dsn> <mode 0-4>");
            std::process::exit(2);
        }
    };

    let stdout = std::io::stdout();
    let mut out = BufWriter::new(stdout.lock());

    if mode == 4 {
        token_dump::dump(&mut out, &path);
        return;
    }

    // `RefWriter.java:26` / `P3T15.java`: the file name with a trailing `.dsn` stripped.
    let design_name = std::path::Path::new(&path)
        .file_name()
        .and_then(|n| n.to_str())
        .unwrap_or_default()
        .strip_suffix(".dsn")
        .unwrap_or_default()
        .to_string();

    let file = std::fs::File::open(&path).unwrap_or_else(|e| panic!("cannot open {path}: {e}"));
    let options = DsnReadOptions::default();
    let result = copper_dsn::read_board(file, None, Some(&design_name), &options);

    let (board, ct, warnings) = match result {
        BoardReadResult::Success {
            board,
            coordinate_transform,
            warnings,
            ..
        }
        | BoardReadResult::OutlineMissing {
            board,
            coordinate_transform,
            warnings,
            ..
        } => match board {
            Some(board) => (*board, coordinate_transform, warnings),
            None => {
                writeln!(out, "RESULT NoBoard").expect("write");
                return;
            }
        },
        // fixed: T4 (#91) — the port's fifth variant. Java has no counterpart, so a fixture that
        // reached it would print a line the Java side cannot, i.e. it would show up as a DIFF in
        // `sweep-p3t15.sh` rather than pass silently.
        BoardReadResult::Partial { diagnostic, .. } => {
            writeln!(out, "RESULT Partial | {diagnostic}").expect("write");
            return;
        }
        BoardReadResult::ParseError { location, detail } => {
            writeln!(out, "RESULT ParseError {location} | {detail}").expect("write");
            return;
        }
        BoardReadResult::IoError(_) => {
            writeln!(out, "RESULT IoError").expect("write");
            return;
        }
    };

    // Java's writers re-derive the transform from `board.communication`; this port takes it
    // explicitly (Task 10, controller ruling A), and `Structure.createBoard`'s is the one that
    // round-trips a file's coordinates unchanged.
    let ct = ct.unwrap_or_else(|| panic!("{path} produced no coordinate transform"));

    match mode {
        0 => dump_items(&mut out, &board, &warnings),
        1 => copper_dsn::dsn_writer::write(&board, &ct, &mut out, &design_name, false).expect("write"),
        2 => copper_dsn::ses_writer::write(&board, &ct, &mut out, &design_name).expect("write"),
        3 => copper_dsn::rules_writer::write(&board, &ct, None, &mut out, &design_name).expect("write"),
        _ => {
            eprintln!("unknown mode: {mode}");
            std::process::exit(2);
        }
    }
}

/// `P3T15.dumpItems`.
fn dump_items(out: &mut impl Write, board: &Board, warnings: &[String]) {
    writeln!(out, "layers {}", board.get_layer_count()).expect("write");
    let mut ids = board.items_in_board_order();
    ids.sort();
    let mut count = 0usize;
    for id in ids {
        let item = board.get_item(id).expect("item in board order exists");
        writeln!(out, "item {} {}", id.0, describe(item, board)).expect("write");
        count += 1;
    }
    writeln!(out, "itemcount {count}").expect("write");
    for w in warnings {
        writeln!(out, "warning {w}").expect("write");
    }
}

/// `P3T15.describe`. `kind` is Java's `getClass().getSimpleName()`.
fn describe(item: &Item, board: &Board) -> String {
    let ctx = board.ctx();
    let kind = match item {
        // The Rust variant is `Trace`; the only concrete Java subclass is `PolylineTrace`.
        Item::Trace(_) => "PolylineTrace",
        Item::Via(_) => "Via",
        Item::Pin(_) => "Pin",
        Item::ObstacleArea(_) => "ObstacleArea",
        Item::ConductionArea(_) => "ConductionArea",
        Item::ViaObstacleArea(_) => "ViaObstacleArea",
        Item::ComponentObstacleArea(_) => "ComponentObstacleArea",
        Item::ComponentOutline(_) => "ComponentOutline",
        Item::BoardOutline(_) => "BoardOutline",
    };
    let mut nets = String::new();
    for i in 0..item.net_count() {
        nets.push_str(&format!("{},", item.get_net_number(i)));
    }
    let box_ = item.bounding_box(&ctx);
    format!(
        "{kind} layers={}..{} nets=[{nets}] cl={} fixed={} cmp={} bbox=({},{},{},{}) tiles={}",
        item.first_layer(&ctx),
        item.last_layer(&ctx),
        item.clearance_class(),
        fixed_state_name(item),
        item.component_id(),
        box_.ll.x,
        box_.ll.y,
        box_.ur.x,
        box_.ur.y,
        item.tile_shape_count(&ctx),
    )
}

/// `FixedState.name()` — the Java enum constant's spelling.
fn fixed_state_name(item: &Item) -> &'static str {
    use copper_board::structure::layer::FixedState;
    match item.get_fixed_state() {
        FixedState::Unfixed => "UNFIXED",
        FixedState::ShoveFixed => "SHOVE_FIXED",
        FixedState::UserFixed => "USER_FIXED",
        FixedState::SystemFixed => "SYSTEM_FIXED",
    }
}
